ygen/gogen.go

```diff
@@ -515,7 +516,23 @@ type {{ $intfName }}_{{ $typeName }} struct {
 // Is_{{ $intfName }} ensures that {{ $intfName }}_{{ $typeName }}
 // implements the {{ $intfName }} interface.
 func (*{{ $intfName }}_{{ $typeName }}) Is_{{ $intfName }}() {}
-{{ end }}`
+{{ end }}
+// To_{{ .Name }} takes an input interface{} and attempts to convert it to a struct
+// which implements the {{ .Name }} union. Returns an error if the interface{} supplied
+// cannot be converted to a type within the union.
+func (t *{{ .ParentReceiver }}) To_{{ .Name }}(i interface{}) ({{ .Name }}, error) {
+	switch v := i.(type) {
+	{{ range $typeName, $type := .Types -}}
+	case {{ $type }}:
+		return &{{ $intfName }}_{{ $typeName }}{v}, nil
+	{{ end -}}
+	default:
+		return nil, fmt.Errorf("cannot convert %v to {{ .Name }}, unknown union type, got: %T, want: 
+		{{- range $typeName, $type := .Types -}}{{ $type }} {{ end -}}
+		", i, i)
```

Comment:
want any of: [type1, type2...] would make the error clearer.

Reply:
Done - required a minor tweak to the input struct based on limitations in Go templating.